lib/auth/recordingmetadata/recordingmetadatav1/recordingmetadata.go

```diff
@@ -0,0 +1,338 @@
+
+			case *apievents.SessionEnd:
+				if !lastActivityTime.IsZero() && e.Time.Sub(lastActivityTime) > inactivityThreshold {
+					addInactivityEvent(lastActivityTime, e.Time)
+				}
+
+				recordThumbnail(e.EndTime)
+
+			case *apievents.SessionJoin:
+				activeUsers[e.User] = e.Time.Sub(startTime)
+
+			case *apievents.SessionLeave:
+				if joinTime, ok := activeUsers[e.User]; ok {
+					metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+						StartOffset: durationpb.New(joinTime),
+						EndOffset:   durationpb.New(e.Time.Sub(startTime)),
+						Event: &pb.SessionRecordingEvent_Join{
+							Join: &pb.SessionRecordingJoinEvent{
+								User: e.User,
+							},
+						},
+					})
+
+					delete(activeUsers, e.User)
+				}
+
+			case *apievents.SessionPrint:
+				if !lastActivityTime.IsZero() && e.Time.Sub(lastActivityTime) > inactivityThreshold {
+					addInactivityEvent(lastActivityTime, e.Time)
+				}
+
+				if _, err := vt.Write(e.Data); err != nil {
+					return trace.Errorf("failed to write data to terminal: %w", err)
+				}
+
+				if sampler.shouldCapture(e.Time) {
+					recordThumbnail(e.Time)
+				}
+
+				lastActivityTime = e.Time
+
+			case *apievents.SessionStart:
+				lastActivityTime = e.Time
+				startTime = e.Time
+
+				cols, rows, err := parseTerminalSize(e.TerminalSize)
+				if err != nil {
+					return trace.Wrap(err, "failed to parse terminal size %q for session %v", e.TerminalSize, sessionID)
+				}
+
+				metadata.ClusterName = e.ClusterName
+
+				metadata.StartCols = int32(cols)
+				metadata.StartRows = int32(rows)
+
+				vt.Resize(cols, rows)
+			}
+
+		case err := <-errors:
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	if lastEvent == nil {
+		return trace.Errorf("could not find any events for session %v", sessionID)
+	}
+
+	// Finish off any remaining activity events
+	for user, userStartOffset := range activeUsers {
+		metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+			StartOffset: durationpb.New(userStartOffset),
+			EndOffset:   durationpb.New(lastEvent.GetTime().Sub(startTime)),
+			Event: &pb.SessionRecordingEvent_Join{
+				Join: &pb.SessionRecordingJoinEvent{
+					User: user,
+				},
+			},
+		})
+	}
+
+	metadata.Duration = durationpb.New(lastEvent.GetTime().Sub(startTime))
+	metadata.StartTime = timestamppb.New(startTime)
+	metadata.EndTime = timestamppb.New(lastEvent.GetTime())
+
+	thumbnails := sampler.result()
+	for _, t := range thumbnails {
+		if t == nil || t.state == nil {
+			continue
+		}
+
+		svg := terminal.VtStateToSvg(t.state)
+
+		thumbnail := &pb.SessionRecordingThumbnail{
+			Svg:         svg,
+			Cols:        int32(t.state.Cols),
+			Rows:        int32(t.state.Rows),
+			CursorX:     int32(t.state.CursorX),
+			CursorY:     int32(t.state.CursorY),
+			StartOffset: durationpb.New(t.startOffset),
+			EndOffset:   durationpb.New(t.endOffset),
+		}
+
+		frames = append(frames, thumbnail)
```

Comment:
instead of accumulating all the frames, should we write them into the buffer as we generate them?